plumbing/object/patch.go

```diff
@@ -0,0 +1,165 @@
+package object
+
+import (
+	"io"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/filemode"
+	fdiff "gopkg.in/src-d/go-git.v4/plumbing/format/diff"
+	"gopkg.in/src-d/go-git.v4/utils/diff"
+
+	dmp "github.com/sergi/go-diff/diffmatchpatch"
+)
+
+func getPatch(message string, changes ...*Change) (fdiff.Patcheable, error) {
+	var filePatches []fdiff.FilePatch
+	for _, c := range changes {
+		from, to, err := c.Files()
+		if err != nil {
+			return nil, err
+		}
+		fromContent, fIsBinary, err := fileContent(from)
+		if err != nil {
+			return nil, err
+		}
+
+		toContent, tIsBinary, err := fileContent(to)
+		if err != nil {
+			return nil, err
+		}
+
+		if fIsBinary || tIsBinary {
+			filePatches = append(filePatches, &textFilePatch{from: c.From, to: c.To})
+
+			continue
+		}
+
+		diffs := diff.Do(fromContent, toContent)
+
+		var chunks []fdiff.Chunk
+		for _, d := range diffs {
+			var op fdiff.Operation
+			switch d.Type {
+			case dmp.DiffEqual:
+				op = fdiff.Equal
+			case dmp.DiffDelete:
+				op = fdiff.Delete
+			case dmp.DiffInsert:
+				op = fdiff.Add
+			}
+
+			chunks = append(chunks, &textChunk{d.Text, op})
+		}
+
+		filePatches = append(filePatches,
+			&textFilePatch{
+				chunks: chunks,
+				from:   c.From,
+				to:     c.To,
+			})
+	}
+
+	return textPatch{message, filePatches}, nil
+}
+
+func fileContent(f *File) (content string, isBinary bool, err error) {
+	if f == nil {
+		return
+	}
+
+	isBinary, err = f.IsBinary()
+	if err != nil || isBinary {
+		return
+	}
+
+	content, err = f.Contents()
+
+	return
+}
+
+type textPatch struct {
+	message     string
+	filePatches []fdiff.FilePatch
+}
+
+func (t textPatch) FilePatches() []fdiff.FilePatch {
```

Comment:
Document all the methods in these files.  A simple `... implements <interface>.` will suffice most of the times.